barretenberg/cpp/src/barretenberg/vm2/constraining/relations/instr_fetching.test.cpp

```diff
@@ -0,0 +1,146 @@
+TEST(InstrFetchingConstrainingTest, EmptyRow)
+{
+    TestTraceContainer trace({
+        { { C::precomputed_clk, 1 } },
+    });
+
+    check_relation<instr_fetching>(trace);
+}
+
+// Basic positive test with a hardcoded bytecode for ADD_8
+TEST(InstrFetchingConstrainingTest, add8WithTraceGen)
+{
+    TestTraceContainer trace;
+    BytecodeTraceBuilder builder;
+    simulation::Instruction add_8_instruction = {
+        .opcode = WireOpCode::ADD_8,
+        .indirect = 3,
+        .operands = { Operand::u8(0x34), Operand::u8(0x35), Operand::u8(0x36) },
+        .size_in_bytes = 5,
+    };
+
+    std::vector<uint8_t> bytecode = { static_cast<uint8_t>(WireOpCode::ADD_8), 0x03, 0x34, 0x35, 0x36 };
+
+    builder.process_instruction_fetching({ { .bytecode_id = 1,
+                                             .pc = 0,
+                                             .instruction = add_8_instruction,
+                                             .bytecode = std::make_shared<std::vector<uint8_t>>(bytecode) } },
+                                         trace);
+
+    EXPECT_EQ(trace.get_num_rows(), 1);
+    check_relation<instr_fetching>(trace);
+}
+
+// Basic positive test with a hardcoded bytecode for ECADD
+TEST(InstrFetchingConstrainingTest, ecaddWithTraceGen)
+{
+    TestTraceContainer trace;
+    BytecodeTraceBuilder builder;
+    simulation::Instruction ecadd_instruction = {
+        .opcode = WireOpCode::ECADD,
+        .indirect = 0x1f1f,
+        .operands = { Operand::u16(0x1279),
+                      Operand::u16(0x127a),
+                      Operand::u16(0x127b),
+                      Operand::u16(0x127c),
+                      Operand::u16(0x127d),
+                      Operand::u16(0x127e),
+                      Operand::u16(0x127f) },
+        .size_in_bytes = 17,
+    };
+
+    std::vector<uint8_t> bytecode = {
+        static_cast<uint8_t>(WireOpCode::ECADD),
+        0x1f,
+        0x1f,
+        0x12,
+        0x79,
+        0x12,
+        0x7a,
+        0x12,
+        0x7b,
+        0x12,
+        0x7c,
+        0x12,
+        0x7d,
+        0x12,
+        0x7e,
+        0x12,
+        0x7f,
+    };
+
+    builder.process_instruction_fetching({ { .bytecode_id = 1,
+                                             .pc = 0,
+                                             .instruction = ecadd_instruction,
+                                             .bytecode = std::make_shared<std::vector<uint8_t>>(bytecode) } },
+                                         trace);
+
+    EXPECT_EQ(trace.get_num_rows(), 1);
+    check_relation<instr_fetching>(trace);
+}
+
+// Positive test for each opcode. We assume that decode instruction is working correctly.
+// It works as long as the relations are not constraining correct range for TAG and or indirect.
+TEST(InstrFetchingConstrainingTest, eachOpcodeWithTraceGen)
+{
+    uint32_t seed = 987137937; // Arbitrary number serving as pseudo-random seed to generate bytes
+
+    auto gen_40_bytes = [&]() {
```

Comment:
bc_decomposition.test.cpp already has a random generator. Please factor out to `vm2/common/testing/fixture.{hpp, cpp}`. Note that I asked the same to @IlyasRidhuan so let's see who gets there first.

You can also just remove the seed if you use that one.